### DISCUSSION TITLE
Google sheet query

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1450] [Ideas & RFCs]
SQL in Google sheets: https://learnsql.com/blog/sql-in-google-sheets-query/

Queries over multiple ranges: https://www.statology.org/google-sheets-query-multiple-ranges/

How to specify entire sheet as range: https://webapps.stackexchange.com/questions/44128/how-to-specify-the-entire-sheet-as-range-in-google-sheets

Google Sheets available data types: https://support.google.com/area120-tables/answer/9904372?hl=en

Currently handled types:
- [x] Number for Integer and Decimal
- [x] String
- [ ] DataTime
  - [x] Formatted String
  - [ ] Timestamp

**Comment by amitu** [comment]
@harshdoesdev 

My first thought was to require `url.replace("https://", "google+sheet://")` or something, but then I realise we are just torturing end users, and we should do some regex search or maybe just a prefix check to decide if its a google sheet URL. 

But Google Sheet is def a case where we will have dozens or hundreds of sheets we want to access in one fastn package, so creating environment variable feels daunting. 

Also can we access private sheets by just the URLs or do we have to do OAuth tokens etc as well? In general I would prefer OAuth stuff so we properly handle access control here.

**Comment by harshdoesdev** [comment]
> Google Sheets are definitely a case where we will have dozens or hundreds of sheets we want to access in one Fastn package, so creating environment variables feels daunting.

Yes, this was one of my concerns as well. We should definitely allow users to directly specify the URL or Google Sheets ID as a header.

> Also, can we access private sheets by just using the URLs, or do we have to use OAuth tokens, etc.? In general, I would prefer OAuth-related solutions so that we can properly handle access control here.

For private sheets, we require OAuth keys or API keys, which the user can generate from the [Google Developer Console](https://console.cloud.google.com/apis/dashboard). The user will also have to enable the Google Sheets API manually.

For public sheets (read-only), users simply need to provide us with the Google Sheet Document ID.

Furthermore, the Google Sheets [Query](https://support.google.com/docs/answer/3093343?hl=en) function is primarily for performing read-only `SELECT` operations for data retrieval. Based on what I have read so far, it does not support INSERT, DELETE, etc.

@amitu

**Reply by amitu** [reply]
> For private sheets, we require OAuth keys or API keys, which the user can generate from the [Google Developer Console](https://console.cloud.google.com/apis/dashboard). The user will also have to enable the Google Sheets API manually.

This is not optimal. We will have to wait for @siddhantk232 to implement OAuth support and Login with Google, with appropriate scope to allow us access to Google Sheet API. Till then let's say private sheet is not allowed.

> Based on what I have read so far, it does not support INSERT, DELETE, etc.

This is fine.

**Reply by amitu** [reply]
>  We should definitely allow users to directly specify the URL or Google Sheets ID as a header.

Sure, let's allow `db: <google-sheet-url>` as a header.

**Reply by harshdoesdev** [reply]
I have created a PR for this: https://github.com/fastn-stack/fastn/pull/1472

**Comment by harshdoesdev** [comment]
[Charts API params](https://developers.google.com/chart/interactive/docs/dev/implementing_data_source#csvdatatable)

**Comment by harshdoesdev** [comment]
https://github.com/orgs/fastn-stack/discussions/1474

**Comment by amitu** [comment]
This is implemented.
